Refactor IntSets

mbauman opened the pull request:
* The element type of `IntSet` is now `Int`.  IntSets now only support integers in the range `1:typemax(Int)`.  This is technically a breaking change on 32-bit systems, where it was previously possible to store some `Int64`s larger than `typemax(Int32)`. Note that all documentation currently says that `IntSet` stores `Int`s, and 32 bit systems currently fail pretty spectacularly for integers larger than 2^36 (a 2GB data structure).  So this breakage is pretty oblique.
* Complete deprecation of stored zeros.
* Complete deprecation of `complement`; removes all support for inverted IntSets
* Refactor internals to rely on a BitVector, allowing the use of highly optimized `map` methods. `IntSet` is now immutable. This significantly improves performance across varying [densities](http://imgur.com/a/uqv8A) and [sizes](http://imgur.com/a/iEgcr). These are compared against a modified Base with deprecation warnings removed for a fairer comparison. Testing code [available here](https://github.com/mbauman/IntSets.jl/tree/b50a7c97abbe9786e33221f723e107e266f31fe4/test).
* Add more tests and organize into testsets.
* Improve hashing; `hash(IntSet([1]))` is now distinct from `hash(IntSet([65]))`

This is a continuation of #10065. Now that complements are fully removed, making IntSet immutable solves the performance issue. I am keeping the name the same within this PR as it vastly simplifies comparisons between the two implementations; the name can later be changed to `IndexSet` if still desired. The naming story is now a bit more complicated since we support offset indices, but a future change could perhaps allow wrapping any `AbstractVector{Bool}` and base the supported `Int`s on those indices. Very few methods depend upon BitArray internals.

I was surprised how easy it was to revive my IntSets package.  The diffstat is deceiving:

```
$ git diff base | diffstat -s
 5 files changed, 183 insertions(+), 291 deletions(-)
$ git diff test | diffstat -s
 1 file changed, 268 insertions(+), 127 deletions(-)
```

mbauman committed:
Refactor IntSets

* Complete deprecation of stored zeros; IntSets now only support integers in the range `1:typemax(Int)`
* Complete deprecation of `complement`; removes all support for inverted IntSets
* Refactor internals to rely on a BitVector, allowing the use of highly optimized `map` methods. `IntSet` is now immutable. This significantly improves performance across varying [densities](http://imgur.com/a/uqv8A) and [sizes](http://imgur.com/a/iEgcr). These are compared against a modified Base with deprecation warnings removed for a fairer comparison. Testing code [available here](https://github.com/mbauman/IntSets.jl/tree/b50a7c97abbe9786e33221f723e107e266f31fe4/test).
* Add more tests and organize into testsets.
* Improve hashing; `hash(IntSet([1]))` is now distinct from `hash(IntSet([65]))`

This is a continuation of #10065. Now that complements are fully removed, making IntSet immutable solves the performance issue. I am keeping the name the same within this PR as it vastly simplifies comparisons between the two implementations; the name can later be changed to `IndexSet` if still desired. The naming story is now a bit more complicated since we support offset indices, but a future change could perhaps allow wrapping any `AbstractVector{Bool}` and base the supported `Int`s on those indices. Very few methods depend upon BitArray internals.

ararslan commented on base/intset.jl:
```diff
@@ -1,339 +1,231 @@
-    IntSet() = new(zeros(UInt32,256>>>5), 256, false)
+immutable IntSet <: AbstractSet{Int}
+    bits::BitVector
+    IntSet() = new(fill!(BitVector(256), false))
```
You should be able to just do `new(falses(256))`, as `falses` returns a `BitVector`.

ararslan commented on base/intset.jl:
```diff
@@ -1,339 +1,231 @@
+# An internal function for setting the inclusion bit for a given integer n >= 0
+@inline function _setint!(s::IntSet, idx::Integer, b::Bool)
+    if idx > length(s.bits)
+        !b && return s # setting a bit to zero outside the set's bits is a no-op
```
`b || return s` avoids the negation, though I'm sure that's so insignificant that it isn't even worth mentioning

ararslan commented on base/intset.jl:
```diff
@@ -1,339 +1,231 @@
+
+# An internal function that takes a pure function `f` and maps across two BitArrays
+# allowing the lengths to be different and altering b1 with the result
+function _matched_map!{F}(f::F, b1::BitArray, b2::BitArray)
```
The `F` parameter shouldn't be necessary I would think?

mbauman replied:
We definitely want these functions to inline and specialize; the type parameter ensures that happens.

iamed2 replied:
It does? I didn't think there would be a difference here. Is there anywhere I can read about this in the docs?

Sorry to butt in, this was just very surprising to me.

mbauman replied:
Hm. There is some sort of difference here that I thought I had measured before… but in this case we're just punting to `map` so it's totally overkill.  I'll ditch the specialization.

ararslan replied:
Yeah I thought the type parameters in cases such as this used to make a difference but at some point something changed so that they didn't matter, but I don't have anything concrete to point to.

nalimilan replied:
@iamed2 AFAIK the rule is that method specialization on function arguments happen only when 1) they are called from the body of the method, or 2) a type parameter is explicitly used. I'm not sure whether that's mentioned in the manual.

iamed2 replied:
@nalimilan Thanks. It makes sense to leave certain compiler implementation details out of the manual if they're expected to change, but if this is a concern for performance it should maybe get a mention?

nalimilan replied:
As I said, it might already be in the manual, so better check that first.

Sacha0 replied:
Ref. #19137 re. specialization on function arguments. Best!

ararslan commented on base/intset.jl:
```diff
@@ -1,339 +1,231 @@
-
-"""
-    symdiff!(s1, s2)
+length(s::IntSet) = sum(s.bits)
```
`countnz` might be more efficient for `BitVector`s than `sum`, but I'm not entirely sure.

carlobaldassi replied:
It's exactly the same, `sum` is just defined as `countnz`.

ararslan commented on base/intset.jl:
```diff
@@ -1,339 +1,231 @@
+    end
+    while i > 0
+        h = hash(bc[i], h)
+	i -= 1
```
This is indented with a single tab but it should be a multiple of 4 spaces

carlobaldassi replied:
There seems to be a tab instead of spaces here.

ararslan commented on test/intset.jl:
```diff
@@ -2,138 +2,279 @@
+    data_in = (1,5,100)
+    s = IntSet(data_in)
+    data_out = collect(s)
+    @test all(map(d->in(d,data_out), data_in))
```
`all` takes a function, so you can do `all(d->in(d, data_out), data_in)`

carlobaldassi commented on base/intset.jl:
```diff
@@ -1,339 +1,231 @@
-        n = Int64(i)
+@inline function in(n::Integer, s::IntSet)
+    if 1 <= n <= length(s.bits)
+        unsafe_getindex(s.bits, n)
```
`@inbounds return s.bits[n]`?

carlobaldassi commented on base/intset.jl:
```diff
@@ -1,339 +1,231 @@
+    to
 end
+eltype(s::IntSet) = Int
+sizehint!(s::IntSet, n::Integer) = (_resize0!(s.bits, n); s)
```
Isn't this going to resize the set no matter what? I think it needs to resize to `isempty(s) ? n : max(n, last(s))`, we don't want to throw away elements larger than `n`.

mbauman replied:
Very good catch.  Fixed and tested.

mbauman committed:
Address review comments; use === where possible

mbauman commented on base/inference.jl:
```diff
@@ -2586,8 +2586,8 @@ function typeinf_frame(frame)
-                min_pc = next(W, Int64(pc) + 1)[1]
-                if min_pc >= W.limit
+                min_pc = next(W, Int64(pc))[2]
+                if done(W, min_pc)
```
Unfortunately yes.  This code depends upon the internal iteration state.

mbauman commented on base/intset.jl:
```diff
@@ -1,339 +1,231 @@
+        newlen = idx + idx>>1 # This operation may overflow; we want saturation
+        _resize0!(s.bits, ifelse(newlen<0, typemax(Int), newlen))
     end
+    unsafe_setindex!(s.bits, b, idx) # Use @inbounds once available
```
Much nicer!

mbauman commented:
Thank you @carlobaldassi and @ararslan.  I've pushed a change that addresses your comments and hopefully makes 32 bit CI a bit happier.  Unfortunately, it does make this an obliquely breaking PR — the eltype is now `Int32` on 32 bit systems, meaning that you can no longer store values bigger than `typemax(Int32)`.

JeffBezanson commented on base/intset.jl:
```diff
@@ -1,339 +1,232 @@
+    if 1 <= n <= length(s.bits)
+        @inbounds b = s.bits[n]
     else
-        n = Int64(ccall(:bitvector_next, UInt64, (Ptr{UInt32}, UInt64, UInt64), s.bits, i, s.limit))
```
I think most of the C functions used in here can be deleted from support/bitvector.* now.

mbauman replied:
I was hoping so myself, but I think it's used by flisp in the parser.

JeffBezanson replied:
Some of the functions are used by flisp, but most aren't. I think everything after bitvector_get in bitvector.c can be deleted.

mbauman replied:
Ah, nice, thanks!

mbauman committed:
Remove no-longer needed C bitvector functions

mbauman committed:
Remove TypeVar in _matched_map!
